server/project/project.go

```diff
@@ -70,5 +125,17 @@ func (s *Server) Delete(ctx context.Context, q *ProjectQuery) (*EmptyResponse, e
 	if !s.enf.EnforceClaims(ctx.Value("claims"), "projects", "delete", fmt.Sprintf("*/%s", q.Name)) {
 		return nil, grpc.ErrPermissionDenied
 	}
+
+	s.projectLock.Lock(q.Name)
+	defer s.projectLock.Unlock(q.Name)
+
+	appsList, err := s.appclientset.ArgoprojV1alpha1().Applications(s.ns).List(metav1.ListOptions{})
```

Comment:
Have you tried ListOptions using a field selector? This may negate the need for FilterByProject()

Reply:
Tried to use filter and seems like it is working only for metadata. I've also changed API to filter by multiple projects and this is not supported for sure :(